### ISSUE TITLE
File history is filtered to list only current and previous commits

### ISSUE THREAD

**Issue by sh2ka** [gitblit/gitblit#916]
I can't see a number of commits for a file in a repository via web interface. But it can be seen using git for windows binary. However, I can see the file in a list of changed files when viewing a commit (that is absent in the file history page) contents in a Commits tab of web interface.

Later I found that when I request history for the file at the point where it was removed then I can see its entire history. It seems that there is some kind of commit date filter to list only previous commits prior to current.

So it appears a difficult task to search entire history for a file in a repository if the file was removed at some point in time.

**Comment by gitblit** [comment]
Thanks for the report and your observations.  Do you have any actionable proposals?

**Comment by sh2ka** [comment]
I think it should show entire history, not prior to selected commit. The selected commit allows to select a file from the list of changed/added/removed files. Otherwise it can mislead the user, because it looks like this is the last commit for the selected file.
Thanks for fast reply.

**Comment by gitblit** [comment]
@sh2ka  this is actually much harder than it appears, at first.  The design you see in Gitblit is how's it always been and the same, actually, as GitHub, Gitweb, and likely others.

You must have a starting point (ref or sha) to define your file history and Gitblit can not automatically assume a file's history only exists on the default branch.  An alternative could be to run through the commit history of all  branches looking for the file, aggregating, and de-duplicating the results but now showing history just got much more expensive and slower.

> So it appears a difficult task to search entire history for a file in a repository if the file was removed at some point in time.

This comment is very true.  This can be solved client-side (http://stackoverflow.com/a/1113140/830200).

    git rev-list -n 1 HEAD -- <file_path>

But you'll notice that the git solution defines a starting point, HEAD, and so disregards any non-merged changes from other branches.

Gitblit could offer a page for showing the history for a manually entered file path to help you discover when something was deleted.  I could see that as useful, but TBH, if you wait for me to write it then you'll be waiting a while.  As a workaround you could manually enter the path to the deleted file in the history url; this will show you the file commit history including when it was deleted - but it still requires setting a starting ref/sha.  Alternatively, you can mine your repo history client-side with git or another tool.

So to summarize, I will not be changing the current design.  The deletion history case can be handled client-side and there is also a (non-elegant) solution server-side.

**Comment by gitblit** [comment]
@sh2ka Thanks for the report.  I don't intend to make any changes.  If you'd like to open a PR with a proposal for improving this, then I'd be happy to review.

**Comment by sh2ka** [comment]
Hi! Thanks for your reply. I used the following command to see entire file history:
`git tree -a`
where `tree` is an alias  in .gitconfig file.
```tree = log --graph --pretty=format:'%Cred%h%Creset -%C(yellow)%d%Creset %s %Cgreen(%cd) %C(bold blue)<%an>%Creset' --abbrev-commit --date=iso```
As you can see I don't specify any ref-s of sha-s.
I mentioned in my comment that when we select a concrete file in a commit (some sha) it would be great to see not only previous commits for it, but the next commits as well if any. It's pretty useful if you want to find changes done by others after your commit. Of course, it relates only files that were removed from the repository. Sorry if I misunderstand something.
Nevertheless, thanks for Gitblit!

**Comment by gitblit** [comment]
Thanks for sharing that alias.  I am regularly impressed by Git's level of flexibility and all the cool tricks that can be performed with it.  JGit doesn't have the same level of convenience, but with some effort I could probably come up something similar.  (Gitblit doesn't use `git`, if you were not aware).

Having said that, Gitblit can't - and shouldn't - cover 100% of all possible git use-cases.  It needs to stay focused on the ~90% of common needs.  Like I wrote previously, your use-case does have utility in Gitblit, but I don't think it fits into the 90% category especially since that history can be mined client-side in multiple ways.
